core/src/solve.jl

```diff
@@ -557,10 +591,12 @@ function formulate_basins!(
     basin::Basin,
     flow::AbstractMatrix,
     storage::AbstractVector,
+    t::Number,
 )::Nothing
     (; node_id, current_level, current_area) = basin
-    current_level = get_tmp(current_level, storage)
-    current_area = get_tmp(current_area, storage)
+    diffvar = isa(t, Dual) ? t : storage
```

Comment:
Could you split this line out into a function with two methods? If the types can be inferred by the compiler, that would eliminate this branch.

You wouldn't need it anymore, but the rest of the code mainly uses `x isa T` syntax rather than `isa(x, T)`.